hail/src/main/scala/is/hail/lir/SplitMethod.scala

```diff
@@ -101,104 +216,347 @@ class SplitMethod(c: Classx[_], m: Method) {
+        val idx = x.children(0)
+        idx.remove()
+        val lidx = splitM.newLocal("switch_index", IntInfo)
+        x.insertBefore(store(lidx, idx))
+        x.setChild(0, load(lidx))
+        val Lreturn = new Block()
+        Lreturn.method = splitM
+        Lreturn.append(returnx(load(lidx)))
+        val newSwitch = switch(splitMReturnValue, x.Ldefault, x.Lcases)
+        var i = 0
+        while (i < x.targetArity()) {
+          val L = x.target(i)
+          if (regionBlocks(L))
+            // null is replaced with throw new SplitUnreachable
+            newSwitch.setTarget(i, null)
+          else
+            x.setTarget(i, Lreturn)
+          i += 1
+        }
+        newL.append(newSwitch)
+      case _: ReturnX =>
+        if (m.returnTypeInfo == UnitInfo)
+          newL.append(returnx())
+        else
+          newL.append(returnx(splitMReturnValue))
+    }
+
+    val b = blockPartitions.find(start)
+    blockSize(b) = newL.approxByteCodeSize()
+    if (splitsReturn)
+      blockSplitsReturn.set(b)
+  }
+
+  // utility class
+  class SizedRange(val start: Int, val end: Int, val size: Int) {
+    override def toString: String = s"$start-$end/$size"
+  }
+
+  def splitSubregions(subr0: mutable.ArrayBuffer[SizedRange]): Int = {
+    var subr = subr0
+
+    var size = subr.iterator.map(_.size).sum
+
+    var changed = true
+    while (changed &&
+      size > SplitMethod.TargetMethodSize) {
 
-    while (x != last) {
-      if (size > SplitMethod.TargetMethodSize) {
-        val newM = c.newMethod(genName("m", "wrapped"), FastIndexedSeq.empty[TypeInfo[_]], UnitInfo)
-        L.method = newM
-        newM.setEntry(L)
-        L.append(returnx())
+      changed = false
 
-        x.insertBefore(methodStmt(INVOKEVIRTUAL, newM, Array(load(m.getParam(0)))))
+      val coalescedsubr = new mutable.ArrayBuffer[SizedRange]()
 
-        L = new Block()
-        size = 0
+      var i = 0
+      while (i < subr.size) {
+        var s = subr(i).size
+        var j = i + 1
+        while (j < subr.size &&
+          subr(j).start == subr(i).end + 1 &&
+          pst.splitBlock(subr(i).end) &&
+          (s + subr(j).size < SplitMethod.TargetMethodSize)) {
+          s += subr(j).size
+          j += 1
+        }
+        coalescedsubr += new SizedRange(subr(i).start, subr(j - 1).end, s)
+        i = j
       }
 
-      size += x.approxByteCodeSize()
-      val n = x.next
-      x.remove()
-      L.append(x)
-      x = n
-    }
+      val sortedsubr = coalescedsubr.sortBy(_.size)
 
-    val newM = c.newMethod(genName("m", "wrapped"), FastIndexedSeq.empty[TypeInfo[_]], returnTI)
-    L.method = newM
-    newM.setEntry(L)
+      val newsubr = mutable.ArrayBuffer[SizedRange]()
 
-    def invokeNewM(): ValueX =
-      methodInsn(INVOKEVIRTUAL, newM, Array(load(m.getParam(0))))
+      i = sortedsubr.length - 1
+      while (i >= 0) {
+        val ri = sortedsubr(i)
+        if (ri.size > 20 &&
```

Comment:
should this be a parameter to include in benchmarking?

Reply:
I don't think so.

This is just to make sure that we don't split something that had just been split, which would go into an infinite loop.  It is paired with this:

```
          splitSlice(ri.start, ri.end)
          val s = blockSize(blockPartitions.find(ri.start))
          assert(s < 20)
```

to make sure split things are sufficiently small that we won't be tempted to resplit them.

I don't think this is a performance issue, but there is potentially a pathological case where splitting could fail to reduce the size of the method: you have a massive region, but is made up of size 15 chlid regions which are all independent with no substructure.  Then nothing would get split, and we might blow the method size limit.  It's actually hard to imagine what such a function might look like, a bunch of switches that all jump to each other.  I'm not quite sure how to split such a function, actually.